### ISSUE TITLE
New attribute comment idea

### ISSUE THREAD

**Issue by chenglou** [facebook/reason#1015]
Dun dun duuuun

Comments aren't part of the AST and it's a pain in the ass to make them interleave correctly. The current algorithm works great, but is undocumented and still slightly buggy (not to mention adds complexity to the code).

Here's a proposal for using a comment attribute, and give it a first-class syntax:

```reason
/* assign 1 to foo */
let foo  = 1 
```

This can desugar to an attribute attached to the let binding!

Benefits over ordinary comments:

- Comments as a first-class AST node, since it's just an agnostic attribute.
- PPXes can now introspect into these "comments" and do potentially neat tricks.
- Comments are now part of the formatting, and part of the width, as they should be.
- You can convert Reason code back to OCaml and the attributes will stay. The code isn't lossy anymore.
- Comments are now clearly attached to a particular node. This means the commenting of the code base is more predictable, e.g.
   
   ```reason
   let foo = 1;
   /* that's the ID being used */
   let bar = 2;
  ```

  Under the new format, we know for sure that the intent of the code is to document the second binding.
- Multi-line comments, no need to worry about manually putting `*` in front anymore. And they'll reformat correctly according to the width.
- ~~Boom, simpler codebase.~~ (In the future)
- Progressive enhancement. If there's a situation we don't cover in the near future, feel free to use the current comments.
- Better documentation generation since you can get the AST and do your stuff, a-la ocamldoc. `[@comment.markdown "## parse parse parse"]`
- Statically analyzable, like ocamldoc's. We can have contributor ppxes that walk through comments of a particular file and assert on things.

Tldr this pushes comments from meta-code into code proper, with all the benefits that comes with that. And it's a non-invasive upgrade.

**Comment by hcarty** [comment]
For doc comments, at least, this comes out of the box with OCaml 4.03.0 and later: https://caml.inria.fr/pub/docs/manual-ocaml/extn.html#sec259

There may be some work there worth piggybacking on.

**Comment by yunxing** [comment]
How would this work with existing OCaml codebase?

**Comment by yunxing** [comment]
I like the idea of putting comments into AST node. It would make comment interleaving much easier. This is the approach that Go and some other languages (Elm?) takes and it worked very well so far.

There are some challenges that are unanswered, though, I guess we can look into what other languages do:
- How do you interleave comments as AST node? Do you write them as parser rules? 
- How could we apply the same logic to parse tree generated by OCaml compiler? Do we always need to have two sets of logic (one is the current logic for comment interleaving, one is the new logic that treats comments as AST nodes)?

**Comment by chenglou** [comment]
@hcarty yeah that's basically what I'm suggesting, exception no conversion necessary: we'd just print `[@comment "foo"]` as `/* foo */` or `// foo`.

@yunxing they're normal attributes so it's just a matter of printing them in the nice location right? Also this can stay small in scope for now. We can keep the same logic for ocaml generated parse tree.

**Comment by hcarty** [comment]
I may be missing something - your examples show the current Reason comment syntax, correct?  Are you going to automatically parse all existing comments into attributes?  Or is there some other syntax than `/* ... */` required to get a comment acting as an attribute?

**Comment by chenglou** [comment]
Whatever exists right now stay the same. I'm suggesting we now print `[@comment "foo"]` as `# foo #` or whatever "comment" syntax you'd like

**Comment by hcarty** [comment]
Ah, cool.  For what it's worth I love the idea!  Making comment formatting more sane is good for everyone.

**Comment by yunxing** [comment]
@chenglou hmm maybe I'm missing something. I still think you need to make dramatic changes to the grammar in order to support putting attributes at arbitrary locations?

**Comment by chenglou** [comment]
We're not gonna put attributes at arbitrary location! =D
We're gonna format comments as deterministically as we do with any other node right now. Got an example though?

**Comment by yunxing** [comment]
@chenglou 

How do we do:
```
let a = /* this is 1 */ 1; 
```
?

**Comment by hcarty** [comment]
One option would be to give the comment attribute a `before`/`after` hint.

```
let a = /* before the 1 */ 1 /* after the 1 */;
```
would be
```
let a = 1 [@comment.before "before the 1"] [@comment.after "after the 1"];
```

**Comment by chenglou** [comment]
@yunxing @hcarty you either will be forced to do that, or will be forced to place it after. Let's say here you wanna place it after 1 but don't know this convention, you can do what we do for precedences currently:

```reason
let a = (# this is 1 # 1)
```

then refmt, and it should refmt to

```reason
let a = 1 # this is 1 #
```

**Comment by yunxing** [comment]
@chenglou I'm still thinking about this approach. Do we need to fork OCaml's parser? 

Currently, OCaml's parser has the type of 
```
source -> (AST, list comments)
```

I think what you are envisioning is the type:
```
source -> AST /* with comments inside */
```

**Comment by bsansouci** [comment]
@chenglou I really like this idea! I think we cannot just change the pretty printer though, because if you write `[@comment ...]` and run `refmt`, you're now left with `(# ... #) ` (whatever syntax you want) which won't parse anymore (right?). 

As @hcarty pointed out, it seems like ocaml already turns comments starting with `**` into attributes. Did reason break this behavior? Why can't we use that for now? 

@yunxing I think cheng's suggesting that we don't need to touch any of the current comment interleaving logic. All that the compiler toolchain sees is that the file has no comments but a bunch of attributes.

**Comment by chenglou** [comment]
@bsansouci you'd parse `(# ... #)` into the attribute

**Comment by hcarty** [comment]
It may be worth pushing some of this work upstream to avoid forking the OCaml parser.  It could be nice in general to have, for example, floating comments generate ephemeral AST nodes which are dropped on printing:

```ocaml
(* x comment above *)
let x = 1
(* x comment below *)

(* A floating comment *)

let y = 1
```
parsing to an AST equivalent to:
```ocaml
let x = 1 [@ocaml.comment.before "x comment above"] [@ocaml.comment.after "x comment below"]

let () = () [@ocaml.ephemeral] [@ocaml.comment "A floating comment"]

let y = 1
```

**Comment by chenglou** [comment]
Btw, are we clear that `(# ... #)` is _not_ a comment thing, it's just sugar for an attribute? Like `{"foo": bar}` and `<Baz />`?

@hcarty it's beside the point, but I feel we should have one way of placing comments in these cases, because otherwise this is ambiguous:

```ocaml
(# x comment above #)
let x = 1

(# x comment below #)
(# A floating comment #)

let y = 1
```

It's really unpredictable to write and suddenly have some parts of your code's comments attached to another node. So I'd say that should probably parse into:

```ocaml
let x = 1 [@ocaml.comment "x comment above"]

let y = 1 [@ocaml.comment "x comment below"] [@ocaml.comment "A floating comment"]
```

Floating comments could work too, if you treat the space between it and the let binding below as significant. But in that case it'd be:

```ocaml
let x = 1 [@ocaml.comment "x comment above"]

let () = () [@ocaml.ephemeral] [@ocaml.comment "x comment below"]
let () = () [@ocaml.ephemeral] [@ocaml.comment "A floating comment"]

let y = 1
```

I'm just arguing about some implementation details here. I feel like we should keep this small for now in case it doesn't work out. No floating comments + comments above bindings is easier to achieve. To get the same intention as your snippet:

Reason: 

```reason
(# x comment above #)
let x = 1;
/* x comment below */

/* A floating comment */

let y = 1;
```

OCaml:

```ocaml
let x = 1 [@ocaml.comment "x comment above"]
/* x comment below */

/* A floating comment */

let y = 1
```

**Comment by hcarty** [comment]
@chenglou I wasn't clear on the sugar-for-an-attribute thing but am now!

I think keeping things simple makes sense.  But how would rearranging comments be avoided in the simple case?
```
(# Here we are before we do a thing #)
let result = do_something_with_side_effects args;
(# That thing with side effects is definitely done now #)

log result;
(# The result has been logged #)
```
Bad comment usage perhaps, but if the comments are rearranged then they become misleading.

**Comment by chenglou** [comment]
Right, and assuming we implement this the easy way, you'd do:

```reason
(# Here we are before we do a thing #)
let result = do_something_with_side_effects args;

(* That thing with side effects is definitely done now *)

log result;
(* The result has been logged *)
```

So the first one is a comment attribute, the rest are "escape hatches" that are just normal comments you'd currently put. I'd say aside from adding another comment syntax, the above snippet is strictly better than three normal comments?

Though even under the assumed implementation constraints, I feel this isn't too bad either:

```reason
(# Here we are before we do a thing #)
let result = do_something_with_side_effects args (# That thing with side effects is definitely done after this call #)

log result (# The result is logged here #);
```

Constrains/formalizes your comment vocabulary a bit, which might even be beneficial (predictable).

**Comment by hcarty** [comment]
So refmt would convert otherwise unattached attribute comments into vanilla comments in your first example, correct? That seems fine.

The second example seems good too but how do you track which comment goes before and which goes after the expression? Is the second example meant to represent the more complex case of before/after comments?

**Comment by chenglou** [comment]
Let's say, in the simpler implementation, no comment can go after (as in, onto a new line). I've updated the previous snippet. I don't think having such constraint is a deal breaker right now, especially when we still have ordinary comments.

For future-proofing, we can desugar `(# foo #)` to `[@ocaml.comment.before "foo"]`, if we add an `[ocaml.comment.after]` later on. Notice that because of such constraint, I've tweaked the wording in my comments. This makes comment reading a bit more predictable too.

**Comment by jaredly** [comment]
commends are now part of the AST (as attributes)
